numpy/lib/function_base.py

```diff
@@ -390,6 +390,32 @@ def iterable(y):
     return True
 
 
+def _weights_are_valid(weights, a, axis):
+    """Validate weights array.
+    
+    We assume, weights is not None.
+    """
+    wgt = np.asanyarray(weights)
+
+    # Sanity checks
+    if a.shape != wgt.shape:
+        if axis is None:
+            raise TypeError(
+                "Axis must be specified when shapes of a and weights "
+                "differ.")
+        if wgt.ndim != 1:
+            raise TypeError(
+                "1D weights expected when shapes of a and weights differ.")
+        if wgt.shape[0] != a.shape[axis]:
+            raise ValueError(
+                "Length of weights not compatible with specified axis.")
+
+        # setup wgt to broadcast along axis
+        wgt = np.broadcast_to(wgt, (a.ndim-1)*(1,) + wgt.shape)
+        wgt = wgt.swapaxes(-1, axis)
+    return wgt
```

Comment:
Interesting logic, but prior art in either case.

Reply:
Why interesting? I just shifted these lines of code out of `average` for reuse in `quantile`.

Reply:
The logic tries to add convenience for an N-D input with with axis and 1-D (or other input) and I find that very dubious guessing of user intent with relatively little gain (it isn't hard to make sure weights and data are properly broadcast/aligned) and potential for confusion.
(The potential for terrible bugs where you expect swapping of weights to happen but it doesn't because shapes happen to be identical seems possible but unlikely.)

Reply:
Again, that code is not new or from me. I just took these lines out of `np.average` for reuse in `np.quantile`.